analyzers/copyslicemap/copy_slice_map.go

```diff
@@ -0,0 +1,100 @@
+// Package copyslicemap defines an Analyzer that checks for
+// uses of copy of slice & map.
+// https://github.com/uber-go/guide/blob/master/style.md#copy-slices-and-maps-at-boundaries
+package copyslicemap
+
+import (
+	"go/ast"
+
+	"golang.org/x/tools/go/analysis"
+)
+
+const Doc = `check for uses of copy of slice & map
+
+This checker reports usages of slice & map that are directly copied. 
+Slices and maps contain pointers to the underlying data so be wary 
+of scenarios when they need to be copied..
+`
+
+var Analyzer = &analysis.Analyzer{
+	Name: "copyslicemap",
+	Doc:  Doc,
+	Run:  run,
+}
+
+func run(pass *analysis.Pass) (interface{}, error) {
+	inspector := newInspector(pass)
+	for _, file := range pass.Files {
+		for _, inspect := range inspector.collections {
+			ast.Inspect(file, inspect)
+		}
+	}
+	return nil, nil
+}
+
+type copySliceMapInspector struct {
+	pass        *analysis.Pass
+	collections []func(node ast.Node) bool
+}
+
+func newInspector(pass *analysis.Pass) *copySliceMapInspector {
+	inspectors := &copySliceMapInspector{
+		pass: pass,
+	}
+	inspectors.collections = []func(ast.Node) bool{
+		inspectors.inspectAssignStatement,
+	}
+	return inspectors
+}
+
+func (i *copySliceMapInspector) inspectAssignStatement(node ast.Node) bool {
+	assignStmt, ok := node.(*ast.AssignStmt)
+	if !ok {
+		return true
+	}
+
+	typ := checkType(assignStmt.Lhs)
+	if typ == "" {
+		return true
+	}
+
+	typ = checkType(assignStmt.Rhs)
+	if typ == "" {
+		return true
+	}
+
```

Comment:
Have an equality type check for Lhs and Rhs? What happens if the direct type contains an interface type but under the hood its an Array or Map?